File: Objective-C/06-ViewControllers/HypnoNerd/HypnoNerd/HypnosisView.m
//
//  HypnosisView.m
//  Hypnosister
//
//  Created by luckytantanfu on 8/9/15.
//  Copyright (c) 2015 futantan. All rights reserved.
//

#import "HypnosisView.h"

@interface HypnosisView ()
@property (strong, nonatomic) UIColor *circleColor;
@end

@implementation HypnosisView

- (instancetype)initWithFrame:(CGRect)frame {
    self = [super initWithFrame:frame];
    if (self) {
        self.backgroundColor = [UIColor clearColor];
        self.circleColor = [UIColor lightGrayColor];
    }

    return self;
}

- (void)setCircleColor:(UIColor *)circleColor {
    _circleColor = circleColor;
    [self setNeedsDisplay];
}


- (void)touchesBegan:(NSSet *)touches withEvent:(UIEvent *)event {
    NSLog(@"%@ was touched", self);

    float red = (float) ((arc4random() % 100) / 100.0);
    float green = (float) ((arc4random() % 100) / 100.0);
    float blue = (float) ((arc4random() % 100) / 100.0);

    UIColor *randomColor = [UIColor colorWithRed:red green:green blue:blue alpha:1.0];
    self.circleColor = randomColor;
}

- (void)drawRect:(CGRect)rect {
    CGRect bounds = self.bounds;

    CGPoint center;
    center.x = (CGFloat) (bounds.origin.x + bounds.size.width / 2.0);
    center.y = (CGFloat) (bounds.origin.y + bounds.size.height / 2.0);

    float maxRadius = (float) (hypot(bounds.size.width, bounds.size.height) / 2.0);

    UIBezierPath *path = [[UIBezierPath alloc] init];

    for (float currentRadius = maxRadius; currentRadius > 0; currentRadius -= 20) {
        [path moveToPoint:CGPointMake(center.x + currentRadius, center.y)];
        [path addArcWithCenter:center
                        radius:currentRadius
                    startAngle:0.0
                      endAngle:(CGFloat) (M_PI * 2.0)
                     clockwise:YES];
    }

    path.lineWidth = 10;
    [self.circleColor setStroke];

    [path stroke];
}

@end

File: Swift/05-RedrawingAndUIScrollView/Hypnosister/Hypnosister/HypnosisView.swift
//
//  HypnosisView.swift
//  Hypnosister
//
//  Created by luckytantanfu on 8/10/15.
//  Copyright (c) 2015 futantan. All rights reserved.
//

import UIKit

class HypnosisView: UIView {
    
    private var circleColor: UIColor {
        didSet {
            self.setNeedsDisplay()
        }
    }
    
    override init(frame: CGRect) {
        circleColor = UIColor.lightGrayColor()
        super.init(frame: frame)
        self.backgroundColor = UIColor.clearColor()
    }
    
    required init(coder aDecoder: NSCoder) {
        circleColor = UIColor.lightGrayColor()
        super.init(coder: aDecoder)
    }
    
    override func touchesBegan(touches: Set<NSObject>, withEvent event: UIEvent) {
        let red = CGFloat(Float(arc4random() % 100) / 100.0)
        let green = CGFloat(Float(arc4random() % 100) / 100)
        let blue = CGFloat(Float(arc4random() % 100) / 100)
        
        let randomColor = UIColor(red: red, green: green, blue: blue, alpha: 1.0)
        
        circleColor = randomColor
    }

    override func drawRect(rect: CGRect) {
        let bounds = self.bounds
        let center = CGPoint(x: bounds.origin.x + bounds.size.width / 2.0, y: bounds.origin.y + bounds.size.height / 2.0)
        let maxRadius = hypot(bounds.size.width, bounds.size.height)
        
        let path = UIBezierPath()
        
        for (var currentRadius = maxRadius; currentRadius > 0; currentRadius = currentRadius - CGFloat(20)) {
            path.moveToPoint(CGPointMake(bounds.width/2.0 + currentRadius, bounds.height/2.0))
            path.addArcWithCenter(center, radius: currentRadius, startAngle: CGFloat(0.0), endAngle: CGFloat(M_PI * 2.0), clockwise: true)
        }
        
        path.lineWidth = 10
        circleColor.setStroke()
        path.stroke()
        
    }
}
